### PR TITLE
Arktos runtime refactor: simplify the northbound imageService API

### PR DESCRIPTION
**Pull request by yb01**
**What type of PR is this?**
> /kind bug

**What this PR does / why we need it**:
This PR simplifies the north bound API impl and removed the internal map for pod-imageService cache mantained by the KuberuntimeManager. it added an extra field to the existing kubecontainer.ImageSpec to rely the info to get the desired image service in PullImage() interface.

This PR also addressed a few TODOs in kuberuntime image service related code.

**Which issue(s) this PR fixes**:
TODOs in code
Partial fix for issue 84.

Fixes #

**Special notes for your reviewer**:

### CODE REVIEW

File: pkg/kubelet/kuberuntime/kuberuntime_image.go [hunk 1]
```diff
@@ -111,42 +104,36 @@ func (m *kubeGenericRuntimeManager) PullImage(image kubecontainer.ImageSpec, pul
 // GetImageRef gets the ID of the image which has already been in
 // the local storage. It returns ("", nil) if the image isn't in the local storage.
 func (m *kubeGenericRuntimeManager) GetImageRef(image kubecontainer.ImageSpec) (string, error) {
+	if image.Pod == nil {
+		klog.Errorf("Pod is not set in imageSpec: %v", image)
+		return "", fmt.Errorf("POD is not set in imageSpec")
+	}
 
 	// Get the image service from the map cache for this image
```
Comment (h-w-chen) [inline pkg/kubelet/kuberuntime/kuberuntime_image.go:112]: is this comment still applicable?

File: pkg/kubelet/container/runtime.go [hunk 2]
```diff
@@ -52,6 +52,7 @@ type Version interface {
 // information about the different image types.
 type ImageSpec struct {
 	Image string
+	Pod *v1.Pod
```
Comment (h-w-chen) [inline pkg/kubelet/container/runtime.go:55]: What is the driving reason to have Pod info in Image spec? Usually image is not pod specific.

File: pkg/kubelet/kuberuntime/kuberuntime_image.go [hunk 3]
```diff
@@ -47,36 +51,27 @@ func (m *kubeGenericRuntimeManager) PullImage(image kubecontainer.ImageSpec, pul
...
-			} else {
-				klog.Infof("Got image ref for image %q: %v", image, imageRef)
-				return imageRef, nil
-			}
+		klog.Infof("Pulling image with image service %v", imageService)
+		imageRef, err := imageService.PullImage(imgSpec, nil, podSandboxConfig)
+		if err != nil {
+			klog.Errorf("Pull image %q failed: %v", img, err)
+		} else {
+			klog.Infof("Pull image ref for image %q: %v", image, imageRef)
```
Comment (h-w-chen) [inline pkg/kubelet/kuberuntime/kuberuntime_image.go:73]: nit: for successful info, we may want to lower the logging priority (confusing; higher level like v5+) to reduce the log file size.

File: pkg/kubelet/kuberuntime/kuberuntime_image.go [hunk 4]
```diff
@@ -111,42 +104,36 @@ func (m *kubeGenericRuntimeManager) PullImage(image kubecontainer.ImageSpec, pul
...
-		status, err := imageService.ImageStatus(&runtimeapi.ImageSpec{Image: image.Image})
-		if err != nil || status == nil {
-			klog.Infof("ImageStatus for image %q failed: %v", image, err)
-		} else {
-			klog.Errorf("Got ImageStatus for image %q: %v", image, status)
-			return status.Id, nil
-		}
-
+	status, err := imageService.ImageStatus(&runtimeapi.ImageSpec{Image: image.Image})
+	if err != nil || status == nil {
```
Comment (h-w-chen) [inline pkg/kubelet/kuberuntime/kuberuntime_image.go:125]: err != nil,   klog.Errorf or Infof?

File: pkg/kubelet/kuberuntime/kuberuntime_image.go [hunk 5]
```diff
@@ -111,42 +104,36 @@ func (m *kubeGenericRuntimeManager) PullImage(image kubecontainer.ImageSpec, pul
...
+		klog.Errorf("Got ImageStatus for image %q: %v", image, status)
+		return status.Id, nil
 	}
 
 	// if not found, return "", nil
 	return "", nil
 }
 
 // ListImages gets all images currently on the machine.
-// TODO: what if the dup image from different imageService ?
```
Comment (h-w-chen) [inline pkg/kubelet/kuberuntime/kuberuntime_image.go:148]: no code change addressing the concerns of the comment to remove. Is the original comment not a real concern, in that case?

File: pkg/kubelet/kuberuntime/kuberuntime_image.go (hunk 4 above)
Comment (h-w-chen) [inline pkg/kubelet/kuberuntime/kuberuntime_image.go:125]: err != nil , Errorf; else Infof?

File: pkg/kubelet/kuberuntime/kuberuntime_image.go (hunk 5 above)
Comment (h-w-chen) [inline pkg/kubelet/kuberuntime/kuberuntime_image.go:148]: no code change addressing the concern recorded by the comment; is it not a real concern? If so, removing the comment makes sense.

Comment (yb01) [reply pkg/kubelet/kuberuntime/kuberuntime_image.go:148]: yes, it should be safe to remove. the Image struct uses ImageId which, in vm runtime, generated from the full image data path from the filesystem. so even with multiple runtime services, the imageID conflict situation should be skipped regardless other runtime service's logic to generated the ID of the image.

File: pkg/kubelet/kuberuntime/kuberuntime_image.go (hunk 4 above)
Comment (yb01) [reply pkg/kubelet/kuberuntime/kuberuntime_image.go:125]: Good catch. my mistake. will fix.

File: pkg/kubelet/kuberuntime/kuberuntime_image.go (hunk 1 above)
Comment (yb01) [reply pkg/kubelet/kuberuntime/kuberuntime_image.go:112]: no. it should be removed. will delete the comment.

File: pkg/kubelet/kuberuntime/kuberuntime_image.go (hunk 3 above)
Comment (yb01) [reply pkg/kubelet/kuberuntime/kuberuntime_image.go:73]: yes, will do. thanks

File: pkg/kubelet/container/runtime.go (hunk 2 above)
Comment (yb01) [reply pkg/kubelet/container/runtime.go:55]: this is a fair ask. for now ( maybe forever ), we won't support mixed workload type so the info used to identify the image service in the API  chain is *pod -- this is mainly to simplify the way to get the desired image service -- using POD info is probably the most straightforward way with the current APIs.

File: pkg/kubelet/kuberuntime/kuberuntime_image.go (hunk 3 above)
Comment (yb01) [reply pkg/kubelet/kuberuntime/kuberuntime_image.go]: fixed.

**Pull request merged**